README.md
[![CocoaPods Compatible](https://img.shields.io/cocoapods/v/JudoKit.svg)](https://img.shields.io/cocoapods/v/JudoKit.svg)
[![Carthage Compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg)](https://github.com/Carthage/Carthage)
[![License](https://img.shields.io/cocoapods/l/JudoKit.svg)](http://http://cocoadocs.org/docsets/Judo)
[![Platform](https://img.shields.io/cocoapods/p/JudoKit.svg)](http://http://cocoadocs.org/docsets/Judo)
[![Twitter](https://img.shields.io/badge/twitter-@JudoPayments-orange.svg)](http://twitter.com/JudoPay)

# Judo Swift SDK for iOS

The judo Swift SDK is a framework for integrating easy, fast and secure payments inside your app with [judo](https://www.judopay.com/). It contains an exhaustive in-app payments and security toolkit that makes integration simple and quick. If you are integrating your app in Objective-C, we highly recommend you to use the [judoKitObjC](https://github.com/judopay/JudoKitObjC) port.

Use our UI components for a seamless user experience for card data capture. Minimise your [PCI scope](https://www.pcisecuritystandards.org/pci_security/completing_self_assessment) with a UI that can be themed or customised to match the look and feel of your app.

## Requirements

Version 7.0.0+ requires Xcode 9 and Swift 4.

Version 6.2.5+ requires Xcode 8 and Swift 3.

Version 6.2.4 is the last version to support Xcode 7.3.1 and Swift 2.2.

## Getting started

#### 1. Integration

Add `import JudoKit` to the top of the file where you want to use the SDK.

If you are integrating using Cocoapods, follow the steps below (if your integration is based on Carthage, then visit https://github.com/JudoPay/JudoKit/wiki/Carthage. If you are integrating the SDK manually then follow the guide here https://github.com/JudoPay/JudoKit/wiki/Manual-integration):

- You can install CocoaPods with the following command:

```bash
$ gem install cocoapods
```

- Add judo to your `Podfile` to integrate it into your Xcode project:

```ruby
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '8.0'
use_frameworks!

pod 'JudoKit', '~> 7.0'
```

- Then run the following command:

```bash
$ pod install
```

- Please make sure to always **use the newly generated `.xcworkspace`** file not not the projects `.xcodeproj` file.

- In your Xcode environment, go to your `Project Navigator` (blue project icon) called `Pods`, select the `JudoKit` target and open the tab called `Build Phases`.
- Add a new `Run Script Phase` and drag it above the `Compile Sources` build phase.
- In the shell script, paste the following line:

```bash
sh "${PODS_ROOT}/DeviceDNA/Framework/strip-frameworks-cocoapods.sh"
```

#### 2. Setup

You can set your token and secret here when initializing the session:

```swift
// initialize the SDK by setting it up with a token and a secret
var judoKit = JudoKit(token: "<TOKEN>", secret: "<SECRET>")
```

To instruct the SDK to communicate with the Sandbox, include the following lines in the ViewController where the payment should be initiated:

```swift
// setting the SDK to Sandbox Mode - once this is set, the SDK wil stay in Sandbox mode until the process is killed
self.judoKit.sandboxed(true)
```

When you are ready to go live you can remove this line.

#### 3. Make a payment

```swift
func paymentOperation() {
    guard let ref = Reference(consumerRef: "<CONSUMER_REFERENCE>") else { return }
    try! self.judoKit.invokePayment(judoId, amount: Amount(decimalNumber: 0.01, currency: currentCurrency), reference: ref, completion: { (response, error) -> () in
        self.dismiss(animated: true, completion: nil)
            
        if let error = error {
            if error.code == .userDidCancel {
                self.dismiss(animated: true, completion: nil)
                return
            }
                
            var errorTitle = "Error"
            if let errorCategory = error.category {
                errorTitle = errorCategory.stringValue()
            }
                
            self.alertController = UIAlertController(title: errorTitle, message: error.message, preferredStyle: .alert)
            self.alertController!.addAction(UIAlertAction(title: "OK", style: .cancel, handler: nil))
            self.dismiss(animated: true, completion:nil)
                
            return // BAIL
        }
            
        if let resp = response, let receipt = resp.items.first {
            self.cardDetails = receipt.cardDetails
            self.paymentToken = receipt.paymentToken()
        }
            
        let sb = UIStoryboard(name: "Main", bundle: nil)
        let viewController = sb.instantiateViewController(withIdentifier: "detailviewcontroller") as! DetailViewController
            
        viewController.response = response
        self.navigationController?.pushViewController(viewController, animated: true)
    })
}
```
**Note:** Please make sure that you are using a unique Consumer Reference for each different consumer.

## Next steps

Judo's Swift SDK supports a range of customization options. For more information on using judo for iOS see our [wiki documentation](https://github.com/JudoPay/JudoKit/wiki/) or [API reference](https://judopay.github.io/JudoKit).


CHANGELOG.md
# Change Log

For the latest SDK changes, please see the [Github release page](https://github.com/Judopay/JudoKit/releases).

## [7.0.0](https://github.com/Judopay/JudoKit/releases/tag/7.0.0)

Swift 4 support! 🎉

## [6.2.21](https://github.com/Judopay/JudoKit/releases/tag/6.2.21)

Certificate pinning 🔑

## [6.2.19](https://github.com/Judopay/JudoKit/releases/tag/6.2.19)

Visa Checkout support 🛒

## [6.2.6](https://github.com/JudoPay/JudoKitObjC/releases/tag/6.2.6)
Released on 2016-11-10

#### Added
- Added in version 2.0.4 of JudoShield.
- Removed call to JudoShield to get current location.

## [6.2.5](https://github.com/JudoPay/JudoKitObjC/releases/tag/6.2.5)
Released on 2016-11-01

#### Added
- SDK now supports Xcode8 and Swift 3

## [6.2.4](https://github.com/JudoPay/JudoKitObjC/releases/tag/6.2.4)
Released on 2016-10-25

#### Added
- Latest version of JudoShield
- Device signals are now encrypted
- Removed redundant Apple Pay sample app

## [6.2.3](https://github.com/JudoPay/JudoKitObjC/releases/tag/6.2.3)
Released on 2016-09-20

#### Added
- Latest version of JudoShield

## [6.2.2](https://github.com/JudoPay/JudoKitObjC/releases/tag/6.2.2)
Released on 2016-09-14

#### Added
- Latest version of JudoShield
- Minor bug fixes.

## [6.2.1](https://github.com/JudoPay/JudoKit/releases/tag/6.2.1)
Released on 2016-06-29

#### Added
- Added a navigation bar title color to the Theme object.

---

## [6.2.0](https://github.com/JudoPay/JudoKit/releases/tag/6.2.0)
Released on 2016-06-02

#### Added
- Amount is now StringLiteralConvertible.

#### Changed
- Injected card details are not masked and can be changed by the user.
- Transaction gets created at initialization to enable adding and removing of custom information.

#### Fixed
- An issue where force unwrapping the bundle was leading to a crash in some rare cases.

---

## [6.1.2](https://github.com/JudoPay/JudoKit/releases/tag/6.1.2)
Released on 2016-05-06

#### Added
- Protocol extension for Card to safely mask important information.
- UI and Integration Tests for duplicate transaction prevention.

#### Changed
- Implemented a protocol as extension in JudoPayInputField.
- Switched casing from 'ID' to 'Id' for correct camel casing.

#### Fixed
- An issue where a secure text field was not found in UI testing.
- An issue where injected card details would not utilize the card number properly.
- An issue where the card type was not parsed correctly.

---

## [6.1.1](https://github.com/JudoPay/JudoKit/releases/tag/6.1.1)
Released on 2016-04-28

#### Changed
- Including DeviceDNA in direct API calls.
- Allowing empty paymentData in Simulator.

#### Fixed
- An issue where generating a new payment reference would fail.

---
## [6.1.0](https://github.com/JudoPay/JudoKit/releases/tag/6.1.0)
Released on 2016-04-21

#### Changed
- Merged Judo-Swift into judoKit.
- Switching from classes to structs.
- More descriptive name for the card security code

#### Fixed
- An issue where presentation on TableViewControllers would fail (Thanks to @andproff)
- An issue where devices set to 12-hour format would fail generating a proper payment reference string

#### Removed
- Theme static accessor.
- Objective-C sample app.

---
## [6.0.2](https://github.com/JudoPay/JudoKit/releases/tag/6.0.2)
Released on 2016-03-30

#### Changed
- Updated podspec.

---
## [6.0.1](https://github.com/JudoPay/JudoKit/releases/tag/6.0.1)
Released on 2016-03-30

#### Changed
- Falling back to a more secure method to detect a valid coordinate when making a transaction using the judo UI.
- Updated Selectors to conform to Swift 3.
- Simplified the security message position updating method.

#### Fixed
- Fixed the security code label for Amex cards (CID instead of CIDV).
- Fixed typos in the header documentation.

---
## [6.0.0](https://github.com/JudoPay/JudoKit/releases/tag/6.0.0)
Released on 2016-03-04

#### Added
- BRL is now a statically accessible from the Currency class.
- Theme object to make it easier to adjust colors, fonts and texts in the JudoKit Payment view.
- JudoPayViewController now needs a judo session to do all transactions.
- NSDate+Extension.swift for easier date handling.
- Hint label text for postcode input field.

Source/Card.swift
//
//  Model.swift
//  Judo
//
//  Copyright (c) 2016 Alternative Payments Ltd
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.

import Foundation


/// Constants that describe the formatting pattern of given Card Networks
let VISAPattern         = "XXXX XXXX XXXX XXXX"
let AMEXPattern         = "XXXX XXXXXX XXXXX"
let CUPPattern          = "XXXXXX XXXXXXXXXXXXX"
let DinersClubPattern   = "XXXX XXXXXX XXXX"


// These should be computed once and then referenced - O(n)
let maestroPrefixes: [String]      = Array([([Int](56...69)).map({ String($0) }), ["50"]].joined())
let dinersClubPrefixes: [String]   = Array([([Int](300...305)).map({ String($0) }), ["36", "38", "39", "309"]].joined())
let instaPaymentPrefixes: [String] = ([Int](637...639)).map({ String($0) })
let JCBPrefixes: [String]          = ([Int](3528...3589)).map({ String($0) })

// Expression was too complex to be executed in one line 😩😭💥
let masterCardPrefixes: [String]   = {
    let m1: [String] = Array(([Int](2221...2720)).map({ String($0) }))
    let m2: [String] = Array(([Int](51...55)).map({ String($0) }))
    return Array([m1, m2].joined())
}()
let discoverPrefixes: [String]     = {
    let d1: [String] = Array(([Int](644...649)).map({ String($0) }))
    let d2: [String] = Array(([Int](622126...622925)).map({ String($0) }))
    return Array([d1, d2, ["65", "6011"]].joined())
}()


/**
 **Card**
 
 Card objects store all the necessary card information for making transactions
*/
public struct Card {
    
    /// The minimum card length constant
    public static let minimumLength = 12
    /// The maximum card length constant
    public static let maximumLength = 19
    
    internal let _number: String
    
    /// The card number should be submitted without any whitespace or non-numeric characters
    public var number: String {
        get {
            return "****"
        }
        set { }
    }
    /// The expiry date should be submitted as MM/YY
    public let expiryDate: String
    
    internal let _securityCode: String
    
    /// CV2 from the credit card, also known as the card verification value (CVV) or security code. It's the 3 or 4 digit number on the back of your credit card
    public var securityCode: String {
        get {
            return "***"
        }
        set { }
    }
    /// The registered add  ress for the card
    public let address: Address?
    /// The start date if the card is a Maestro
    public let startDate: String?
    /// The issue number if the card is a Maestro
    public let issueNumber: String?
    
    
    /**
     Designated initializer for the Card struct
     
     - parameter number:      The card number (long number)
     - parameter expiryDate:  The expiry date of the card
     - parameter cv2:         The security code number of the card
     - parameter address:     The address of the card holder where the account is registered (AVS)
     - parameter startDate:   In case of transacting with maestro cards, the start date (optional)
     - parameter issueNumber: In case of transacting with maestro cards, the issue number (optional)
     
     - returns: a Card object
     */
    public init(number: String, expiryDate: String, securityCode: String, address: Address? = nil, startDate: String? = nil, issueNumber: String? = nil) {
        self._number = number
        self.expiryDate = expiryDate
        self._securityCode = securityCode
        self.address = address
        self.startDate = startDate
        self.issueNumber = issueNumber
    }
    
    /**
     **Card.Configuration**
     
     Card Configuration consists of a Card Network and a given length
    */
    public struct Configuration {
        /// The network of the configuration
        public let cardNetwork: CardNetwork
        /// The length of the card for this configuration
        public let cardLength: Int
        
        /**
         Designated initializer for a card configuration
         
         - parameter cardNetwork: the card network (eg. Visa, MasterCard or AMEX)
         - parameter cardLength:  the length of the card number (eg. 16 or 19 for Maestro cards)
         
         - returns: a Card.Configuration object
         */
        public init(_ cardNetwork: CardNetwork, _ cardLength: Int) {
            self.cardLength = cardLength
            self.cardNetwork = cardNetwork
        }
        
        
        /**
        Helper method to get a pattern string for a certain configuration
        
        - Returns: a given String with the correct pattern
        */
        public func patternString() -> String? {
            switch self.cardNetwork {
            case .visa, .masterCard, .dankort, .jcb, .instaPayment, .discover:
                return VISAPattern
            case .amex:
                return AMEXPattern
            case .chinaUnionPay, .interPayment, .maestro:
                if self.cardLength == 16 {
                    return VISAPattern
                } else if self.cardLength == 19 {
                    return CUPPattern
                }
                break
            case .dinersClub:
                if self.cardLength == 16 {
                    return VISAPattern
                } else if self.cardLength == 14 {
                    return DinersClubPattern
                }
            default:
                return VISAPattern
            }
            return nil
        }
        
        
        /**
        Helper method to get a placeholder string for a certain configuration
        
        - Returns: a given String as a placeholder
        */
        public func placeholderString() -> String? {
            return self.patternString()?.replacingOccurrences(of: "X", with: "0")
        }
    }
    
}

// MARK: - Card Protocol extensions

extension Card: CustomStringConvertible {
    public var description: String {
        return "number: \(self.number), expiryDate: \(self.expiryDate), securityCode: \(self.securityCode), address: \(String(describing: self.address)), startDate: \(String(describing: self.startDate)), issueNumber: \(String(describing: self.issueNumber))"
    }
}

// MARK: - Card.Configurations Protocol extensions
extension Card.Configuration: Comparable { }


/**
 Equals function for two Card.Configuration objects
 
 - parameter x: left hand side Card.Configuration
 - parameter y: right hand side Card.Configuration
 
 - returns: boolean indicating whether the two objects are equal
 */
public func ==(x: Card.Configuration, y: Card.Configuration) -> Bool {
    return x.cardLength == y.cardLength && x.cardNetwork == y.cardNetwork
}


/**
 Lower function for two Card.Configuration objects
 
 - parameter x: left hand side Card.Configuration
 - parameter y: right hand side Card.Configuration
 
 - returns: boolean indicating whether the lhs objects cardLength is lower than the right hand sides cardLength
 */
public func <(x: Card.Configuration, y: Card.Configuration) -> Bool {
    return x.cardLength < y.cardLength
}


/**
 Lower or equal function for two Card.Configuration objects
 
 - parameter x: left hand side Card.Configuration
 - parameter y: right hand side Card.Configuration
 
 - returns: boolean indicating whether the lhs objects cardLength is lower than or equals the right hand sides cardLength
 */
public func <=(x: Card.Configuration, y: Card.Configuration) -> Bool {
    return x.cardLength <= y.cardLength
}


/**
 Greater function for two Card.Configuration objects
 
 - parameter x: left hand side Card.Configuration
 - parameter y: right hand side Card.Configuration
 
 - returns: boolean indicating whether the lhs objects cardLength is greater than the right hand sides cardLength
 */
public func >(x: Card.Configuration, y: Card.Configuration) -> Bool {
    return x.cardLength > y.cardLength
}


/**
 Greater or equal function for two Card.Configuration objects
 
 - parameter x: left hand side Card.Configuration
 - parameter y: right hand side Card.Configuration
 
 - returns: boolean indicating whether the lhs objects cardLength is greater than or equals the right hand sides cardLength
 */
public func >=(x: Card.Configuration, y: Card.Configuration) -> Bool {
    return x.cardLength >= y.cardLength
}

// MARK - Additional card helpers

/**
The CardType enum is a value type for CardNetwork to further identify card types

- Debit:   Debit Card type
- Credit:  Credit Card type
- Unknown: Unknown Card type
*/
public enum CardType: Int {
    /// Debit Card type
    case debit
    /// Credit Card type
    case credit
    /// Unknown Card type
    case unknown
}


/**
The CardNetwork enum depicts the Card Network type of a given Card object

- Visa:             Visa Card Network
- MasterCard:       MasterCard Network
- AMEX:             American Express Card Network
- DinersClub:       Diners Club Network
- Maestro:          Maestro Card Network
- ChinaUnionPay:    China UnionPay Network
- Discover:         Discover Network
- InterPayment:     InterPayment Network
- InstaPayment:     InstaPayment Network
- JCB:              JCB Network
- Dankort:          Dankort Network
- UATP:             UATP Network
- Unknown:          Unknown
*/
public enum CardNetwork: Int64 {
    
    /// Unknown
    case unknown = 0
    /// Visa Card Network
    case visa = 1
    /// MasterCard Network
    case masterCard = 2
    /// Visa Electron Network
    case visaElectron = 3
    /// Switch Network
    case `switch` = 4
    /// Solo Network
    case solo = 5
    /// Laser Network
    case laser = 6
    /// China UnionPay Network
    case chinaUnionPay = 7
    /// American Express Card Network
    case amex = 8
    /// JCB Network
    case jcb = 9
    /// Maestro Card Network
    case maestro = 10
    /// Visa Debit Card Network
    case visaDebit = 11
    /// MasterCard Network
    case masterCardDebit = 12
    /// Visa Purchasing Network
    case visaPurchasing = 13
    /// Discover Network
    case discover = 14
    /// Carnet Network
    case carnet = 15
    /// Carte Bancaire Network
    case carteBancaire = 16
    /// Diners Club Network
    case dinersClub = 17
    /// Elo Network
    case elo = 18
    /// Farmers Card Network
    case farmersCard = 19
    /// Soriana Network
    case soriana = 20
    /// Private Label Card Network
    case privateLabelCard = 21
    /// Q Card Network
    case qCard = 22
    /// Style Network
    case style = 23
    /// True Rewards Network
    case trueRewards = 24
    /// UATP Network
    case uatp = 25
    /// Bank Card Network
    case bankCard = 26
    /// Banamex Costco Network
    case banamex_Costco = 27
    /// InterPayment Network
    case interPayment = 28
    /// InstaPayment Network
    case instaPayment = 29
    /// Dankort Network
    case dankort = 30
    
    
    /**
     The string value of the receiver
     
     - returns: a string describing the receiver
     */
    public func stringValue() -> String {
        switch self {
        case .unknown:
            return "Unknown"
        case .visa, .visaDebit, .visaElectron, .visaPurchasing:
            return "Visa"
        case .masterCard, .masterCardDebit:
            return "MasterCard"
        case .switch:
            return "Switch"
        case .solo:
            return "Solo"
        case .laser:
            return "Laser"
        case .chinaUnionPay:
            return "China UnionPay"
        case .amex:
            return "AMEX"
        case .jcb:
            return "JCB"
        case .maestro:
            return "Maestro"
        case .discover:
            return "Discover"
        case .carnet:
            return "Carnet"
        case .carteBancaire:
            return "CarteBancaire"
        case .dinersClub:
            return "Diners Club"
        case .elo:
            return "Elo"
        case .farmersCard:
            return "FarmersCard"
        case .soriana:
            return "Soriana"
        case .privateLabelCard:
            return "PrivateLabelCard"
        case .qCard:
            return "QCard"
        case .style:
            return "Style"
        case .trueRewards:
            return "TrueRewards"
        case .uatp:
            return "UATP"
        case .bankCard:
            return "BankCard"
        case .banamex_Costco:
            return "BanamexCostco"
        case .interPayment:
            return "InterPayment"
        case .instaPayment:
            return "InstaPayment"
        case .dankort:
            return "Dankort"
        }
    }
    
    
    /**
    The prefixes determine which card network a number belongs to, this method provides an array with one or many prefixes for a given type
    
    - Returns: an Array containing all the possible prefixes for a type
    */
    public func prefixes() -> [String] {
        switch self {
        case .visa, .visaDebit, .visaElectron, .visaPurchasing:
            return ["4"]
        case .masterCard, .masterCardDebit:
            return masterCardPrefixes
        case .amex:
            return ["34", "37"]
        case .dinersClub:
            return dinersClubPrefixes
        case .maestro:
            return maestroPrefixes
        case .chinaUnionPay:
            return ["62"]
        case .discover:
            return discoverPrefixes
        case .interPayment:
            return ["636"]
        case .instaPayment:
            return instaPaymentPrefixes
        case .jcb:
            return JCBPrefixes
        case .dankort:
            return ["5019"]
        case .uatp:
            return ["1"]
        default:
            return []
        }
    }
    
    
    /**
     The card network type for a given card number string and constrained to a set of networks
     
     - parameter string:   the card number as a string
     - parameter networks: the networks allowed for detection
     
     - returns: a CardNetwork if the prefix matches a given set of CardNetworks or CardNetwork.Unknown
     */
    public static func networkForString(_ string: String, constrainedToNetworks networks: [CardNetwork]) -> CardNetwork {
        let result = networks.filter({ $0.prefixes().filter({ string.hasPrefix($0) }).count > 0 })
        //Case when maestro and china union pay are returned by the filter
        if result.count == 2 {
            return result[1]
        }
        if result.count == 1 {
            return result[0]
        }
        return CardNetwork.unknown
    }
    
    
    /**
     The card network type for a given card number
     
     - parameter string: the card number as a string
     
     - returns: a CardNetwork if the prefix matches any CardNetwork prefix
     */
    public static func networkForString(_ string: String) -> CardNetwork {
        let allNetworks: [CardNetwork] = [.visa, .masterCard, .amex, .dinersClub, .maestro, .chinaUnionPay, .discover, .interPayment, .instaPayment, .jcb, .dankort, .uatp]
        return self.networkForString(string, constrainedToNetworks: allNetworks)
    }
    

    /**
    Security code name for a certain card
    
    - Returns: a String for the title of a certain security code
    */
    public func securityCodeTitle() -> String {
        switch self {
        case .visa:
            return "CVV2"
        case .masterCard:
            return "CVC2"
        case .amex:
            return "CID"
        case .chinaUnionPay:
            return "CVN2"
        case .discover:
            return "CID"
        case .jcb:
            return "CAV2"
        case .unknown:
            return "CVV"
        default:
            return "CVV"
        }
    }
    
    
    /**
    Security code length for a card type
    
    - Returns: an Int with the code length
    */
    public func securityCodeLength() -> Int {
        switch self {
        case .amex:
            return 4
        default:
            return 3
        }
    }
    
}


// MARK - CardDetails

/**
 **CardDetails**
 
 The CardDetails object stores information that is returned from a successful payment or pre-auth
 
 This class also implements the `NSCoding` protocol to enable serialization for persistency
*/
open class CardDetails: NSObject, NSCoding {
    /// The last four digits of the card used for this transaction
    open var cardLastFour: String?
    /// Expiry date of the card used for this transaction formatted as a two digit month and year i.e. MM/YY
    open let endDate: String?
    /// Can be used to charge future payments against this card
    open let cardToken: String?
    
    fileprivate var _cardNetwork: CardNetwork?
    /// The computed card network
    open var cardNetwork: CardNetwork? {
        get {
            if let computedCardNetwork = self._cardNumber?.cardNetwork() , self._cardNetwork == .unknown && self._cardNumber != nil {
                self._cardNetwork = computedCardNetwork
            }
            return self._cardNetwork
        }
        set {
            self._cardNetwork = newValue
        }
    }
    
    internal let _cardNumber: String?
    
    /// The card number if available
    open var cardNumber: String? {
        get {
            return self.formattedLastFour()
        }
        set {}
    }
    
    open var isCardNumberValid: Bool {
        get {
            return self._cardNumber?.isCardNumberValid() ?? false
        }
        set { }
    }
    
    /// Description string for print functions
    override open var description: String {
        let formattedLastFour = self.formattedLastFour() ?? "N/A"
        let formattedEndDate = self.formattedEndDate() ?? "N/A"
        let cardNetworkString = self.cardNetwork?.stringValue() ?? "N/A"
        return "cardLastFour: \(formattedLastFour), endDate: \(formattedEndDate), cardNetwork: \(cardNetworkString)"
    }
    
    
    /**
     Convenience initializer that takes in a card number and an expiry date in case a transaction should be made with pre-known information
     
     - parameter cardNumber: a card number as a String
     - parameter endDate:    an enddate in MM/YY format as a String
     
     - returns: a CardDetails object
     */
    public convenience init(cardNumber: String?, expiryMonth: Int?, expiryYear: Int?) {
        var dict = JSONDictionary()
        dict["cardNumber"] = cardNumber as AnyObject?
        
        guard let expiryMonth = expiryMonth, let expiryYear = expiryYear else { self.init(dict); return }
        
        var dateComponents = DateComponents()
        dateComponents.month = expiryMonth
        dateComponents.year = expiryYear
        
        let gregorian = Calendar(identifier: Calendar.Identifier.gregorian)
        
        if let date = gregorian.date(from: dateComponents) {
            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "MM/yy"
            let dateString = dateFormatter.string(from: date)
            dict["endDate"] = dateString as AnyObject?
        }
        
        self.init(dict)
    }
    
    
    /**
     Designated initializer for Card Details
     
     - parameter dict: all parameters as a dictionary
     
     - returns: a CardDetails object
     */
    public init(_ dict: JSONDictionary?) {
        self.cardLastFour = dict?["cardLastfour"] as? String
        self.endDate = dict?["endDate"] as? String
        self.cardToken = dict?["cardToken"] as? String
        self._cardNumber = dict?["cardNumber"] as? String
        super.init()
        if let cardType = dict?["cardType"] as? Int {
            self.cardNetwork = CardNetwork(rawValue: Int64(cardType))
        } else {
            self.cardNetwork = .unknown
        }
    }
    
    
    /**
     Initialise the CardDetails object with a coder
     
     - parameter decoder: the decoder object
     
     - returns: a CardDetails object or nil
     */
    public required init?(coder decoder: NSCoder) {
        let cardLastFour = decoder.decodeObject(forKey: "cardLastFour") as? String?
        let endDate = decoder.decodeObject(forKey: "endDate") as? String?
        let cardToken = decoder.decodeObject(forKey: "cardToken") as? String?
        let cardNetwork = decoder.decodeInt64(forKey: "cardNetwork")
        
        self.cardLastFour = cardLastFour ?? nil
        self.endDate = endDate ?? nil
        self.cardToken = cardToken ?? nil
        self._cardNumber = nil
        super.init()
        self.cardNetwork = CardNetwork(rawValue: Int64(cardNetwork))
    }
    
    
    /**
     Encode the receiver CardDetails object
     
     - parameter aCoder: the Coder
     */
    open func encode(with aCoder: NSCoder) {
        aCoder.encode(self.cardLastFour, forKey: "cardLastFour")
        aCoder.encode(self.endDate, forKey: "endDate")
        aCoder.encode(self.cardToken, forKey: "cardToken")
        if let cardNetwork = self.cardNetwork {
            aCoder.encode(cardNetwork.rawValue, forKey: "cardNetwork")
        } else {
            aCoder.encode(0, forKey: "cardNetwork")
        }
    }
    
    
    /**
     Get a formatted string with the right whitespacing for a certain card type
     
     - returns: a string with the last four digits with the right format
     */
    open func formattedLastFour() -> String? {
        if self.cardLastFour == nil && self._cardNumber == nil {
            return nil
        } else if let cardNumber = self._cardNumber {
            self.cardLastFour = String(cardNumber[cardNumber.index(cardNumber.endIndex, offsetBy: -4)...])
        }
        
        guard let cardNetwork = self.cardNetwork else { return "**** \(cardLastFour!)" }
        
        switch cardNetwork {
        case .amex:
            return "**** ****** *\(cardLastFour!)"
        default:
            return "**** **** **** \(cardLastFour!)"
        }
    }
    
    
    /**
     Get a formatted string with the right slash for a date
     
     - returns: a string with the date as shown on the credit card with the right format
     */
    open func formattedEndDate() -> String? {
        guard let ed = self.endDate else { return nil }
        
        if ed.count == 4 {
            // backend returns the end date without a slash so for UI purposes we have to add it
            let prefix = ed[..<ed.index(ed.startIndex, offsetBy: 2)]
            let suffix = ed[ed.index(ed.endIndex, offsetBy: -2)...]
            return "\(prefix)/\(suffix)"
        } else {
            return ed
        }
    }
    
}

